sdk/core/System.Buffers.Data/src/BinaryData.cs

```diff
@@ -0,0 +1,299 @@
+            if (stream.CanSeek)
+            {
+                long longLength = stream.Length - stream.Position;
+                if (longLength > int.MaxValue)
+                {
+                    throw new ArgumentOutOfRangeException(
+                        nameof(stream),
+                        "Stream length must be less than Int32.MaxValue");
+                }
+                streamLength = (int)longLength;
+            }
+
+            using MemoryStream memoryStream = stream.CanSeek ? new MemoryStream(streamLength) : new MemoryStream();
+            if (async)
+            {
+                await stream.CopyToAsync(memoryStream, CopyToBufferSize, cancellationToken).ConfigureAwait(false);
+            }
+            else
+            {
+                stream.CopyTo(memoryStream);
+            }
+            return new BinaryData(data: memoryStream.GetBuffer().AsMemory(0, (int)memoryStream.Position), ownsBuffer: true);
+        }
+
+        /// <summary>
+        /// Creates a <see cref="BinaryData"/> instance by serializing the provided object using
+        /// the <see cref="JsonSerializer"/>.
+        /// </summary>
+        ///
+        /// <typeparam name="T">The type to use when serializing the data.</typeparam>
+        /// <param name="jsonSerializable">The data to use.</param>
+        /// <param name="options">The options to use when serializing to JSON.</param>
+        ///
+        /// <returns>A value representing the UTF-8 encoding of the JSON representation of <paramref name="jsonSerializable" />.</returns>
+        public static BinaryData FromObjectAsJson<T>(
+            T jsonSerializable,
+#pragma warning disable AZC0014 // Avoid using banned types in public API
+            JsonSerializerOptions? options = default)
+#pragma warning restore AZC0014 // Avoid using banned types in public API
+        {
+            byte[] buffer = JsonSerializer.SerializeToUtf8Bytes(jsonSerializable, typeof(T), options);
+            return new BinaryData(data: buffer, ownsBuffer: true);
+        }
+
+        /// <summary>
+        /// Converts the value of this instance to a string using UTF-8.
+        /// </summary>
+        /// <returns>
+        /// A string from the value of this instance, using UTF-8 to decode the bytes.
+        /// </returns>
+        public override string ToString()
+        {
+            if (MemoryMarshal.TryGetArray(
+                _bytes,
+                out ArraySegment<byte> data))
+            {
+                return Encoding.UTF8.GetString(data.Array, data.Offset, data.Count);
+            }
+            return Encoding.UTF8.GetString(_bytes.ToArray());
+        }
+
+        /// <summary>
+        /// Converts the <see cref="BinaryData"/> to a read-only stream.
+        /// </summary>
+        /// <returns>A stream representing the data.</returns>
+        public Stream ToStream() =>
+            new ReadOnlyMemoryStream(_ownsBuffer ? _bytes : _bytes.ToArray());
```

Comment:
I think we should apply NeverUseThisAPIAttribute to this method and recommend that people create their own stream and just wrap the return value of ToMemory. :-)

Reply:
That would burn a perfectly good API.

Reply:
For what it's worth, I think that the copying/disassociation with the input is a virtue; but if I'm in the minority I'll just retreat to the shadows on this.

Reply:
Will undo.